File: youtubeapi/src/main/java/com/liskovsoft/youtubeapi/next/v2/result/WatchNextResult.kt
package com.liskovsoft.youtubeapi.next.v2.result

data class WatchNextResult(
    val contents: Contents?,
    val transportControls: TransportControls?
) {
    data class Contents(
        val singleColumnWatchNextResults: SingleColumnWatchNextResults?
    ) {
        data class SingleColumnWatchNextResults(
            val pivot: Pivot?,
            val results: Results?,
            val autoplay: Autoplay?
        ) {
            data class Pivot(
                val pivot: Pivot?
            ) {
                data class Pivot(
                    val contents: List<Content?>?
                ) {
                    data class Content(
                        val shelfRenderer: ShelfRenderer?
                    ) {
                        data class ShelfRenderer(
                            val title: Title?,
                            val content: Content?
                        ) {
                            data class Title(
                                val runs: List<Run?>?,
                                val simpleText: String?
                            ) {
                                data class Run(
                                    val text: String?
                                )
                            }

                            data class Content(
                                val horizontalListRenderer: HorizontalListRenderer?
                            ) {
                                data class HorizontalListRenderer(
                                    val items: List<Item?>?
                                ) {
                                    data class Item(
                                        val tileRenderer: TileRenderer?,
                                        val gridVideoRenderer: GridVideoRenderer?,
                                        val pivotVideoRenderer: PivotVideoRenderer?
                                    ) {
                                        data class TileRenderer(
                                            val metadata: Metadata?
                                        ) {
                                            data class Metadata(
                                                val tileMetadataRenderer: TileMetadataRenderer?
                                            ) {
                                                data class TileMetadataRenderer(
                                                    val title: Title?
                                                ) {
                                                    data class Title(
                                                        val runs: List<Run?>?,
                                                        val simpleText: String?
                                                    ) {
                                                        data class Run(
                                                            val text: String?
                                                        )
                                                    }
                                                }
                                            }
                                        }

                                        data class GridVideoRenderer(
                                            val title: Title?
                                        ) {
                                            data class Title(
                                                val runs: List<Run?>?,
                                                val simpleText: String?
                                            ) {
                                                data class Run(
                                                    val text: String?
                                                )
                                            }
                                        }

                                        data class PivotVideoRenderer(
                                            val title: Title?
                                        ) {
                                            data class Title(
                                                val runs: List<Run?>?,
                                                val simpleText: String?
                                            ) {
                                                data class Run(
                                                    val text: String?
                                                )
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }

            data class Results(
                val results: Results?
            ) {
                data class Results(
                    val contents: List<Content?>?
                ) {
                    data class Content(
                        val itemSectionRenderer: ItemSectionRenderer?
                    ) {
                        data class ItemSectionRenderer(
                            val contents: List<Content?>?
                        ) {
                            data class Content(
                                val videoMetadataRenderer: VideoMetadataRenderer?,
                                val musicWatchMetadataRenderer: MusicWatchMetadataRenderer?
                            ) {
                                data class VideoMetadataRenderer(
                                    val owner: Owner?
                                ) {
                                    data class Owner(
                                        val videoOwnerRenderer: VideoOwnerRenderer?
                                    ) {
                                        class VideoOwnerRenderer
                                    }
                                }

                                class MusicWatchMetadataRenderer
                            }
                        }
                    }
                }
            }

            data class Autoplay(
                val autoplay: Autoplay?
            ) {
                data class Autoplay(
                    val sets: List<Set?>?,
                    val replayVideoRenderer: ReplayVideoRenderer?
                ) {
                    data class Set(
                        val nextVideoRenderer: NextVideoRenderer?
                    ) {
                        data class NextVideoRenderer(
                            val maybeHistoryEndpointRenderer: MaybeHistoryEndpointRenderer?,
                            val autoplayEndpointRenderer: AutoplayEndpointRenderer?
                        ) {
                            class MaybeHistoryEndpointRenderer

                            class AutoplayEndpointRenderer
                        }
                    }

                    class ReplayVideoRenderer
                }
            }
        }
    }

    data class TransportControls(
        val transportControlsRenderer: TransportControlsRenderer?
    ) {
        class TransportControlsRenderer
    }
}

File: youtubeapi/src/main/java/com/liskovsoft/youtubeapi/next/v2/result/WatchNextResult.js
let buttonState = {
    subscribeButton: {
        toggleButtonRenderer: {
            isToggled: false
        }
    },
    likeButton: {
        toggleButtonRenderer: {
            isToggled: false
        }
    },
    dislikeButton: {
        toggleButtonRenderer: {
            isToggled: false
        }
    },
    channelButton: {
        videoOwnerRenderer: {
            navigationEndpoint: {
                browseEndpoint: {
                    browseId: ""
                }
            },
            thumbnail: thumbnailItem
        }
    }
};

let itemWrapper = {
    // TileItem
    tileRenderer: tileItem,
    // VideoItem
    gridVideoRenderer: videoItem,
    // SuggestedVideoItem
    pivotVideoRenderer: videoItem,
    // MusicItem
    tvMusicVideoRenderer: musicItem,
    // RadioItem
    gridRadioRenderer: radioItem,
    pivotRadioRenderer: radioItem,
    // ChannelItem
    gridChannelRenderer: channelItem,
    pivotChannelRenderer: channelItem,
    // PlaylistItem
    gridPlaylistRenderer: playlistItem,
    pivotPlaylistRenderer: playlistItem
};

let nextVideoItem = {

};

let watchNextResult = {
    contents: {
        singleColumnWatchNextResults: {
            // SuggestedSections
            // $.contents.singleColumnWatchNextResults.pivot.pivot.contents[*].shelfRenderer
            // $.title.runs[0].text, $.title.simpleText
            pivot: {
                pivot: {
                    contents: [
                        {
                            shelfRenderer: {
                                title: textItem,
                                // ItemWrappers
                                // $.content.horizontalListRenderer.items[*]
                                content: {
                                    horizontalListRenderer: {
                                        items: [
                                            itemWrapper
                                        ]
                                    }
                                }
                            }
                        }
                    ]
                }
            },
            // VideoMetadata and VideoOwner
            // $.contents.singleColumnWatchNextResults.results.results.contents[0].itemSectionRenderer.contents[0].videoMetadataRenderer
            // $.contents.singleColumnWatchNextResults.results.results.contents[0].itemSectionRenderer.contents[0].musicWatchMetadataRenderer
            results: {
                results: {
                    contents: [
                        {
                            itemSectionRenderer: {
                                contents: [
                                    {
                                        videoMetadataRenderer: videoMetadataItem,
                                        musicWatchMetadataRenderer: videoMetadataItem
                                    }
                                ]
                            }
                        }
                    ]
                }
            },
            // NextVideo
            // VideoDetails
            // ReplayItem
            // $.contents.singleColumnWatchNextResults.autoplay.autoplay.sets[0].nextVideoRenderer.maybeHistoryEndpointRenderer
            // $.contents.singleColumnWatchNextResults.autoplay.autoplay.sets[0].nextVideoRenderer.autoplayEndpointRenderer
            // $.contents.singleColumnWatchNextResults.autoplay.autoplay.sets.replayVideoRenderer.pivotVideoRenderer ????
            // $.contents.singleColumnWatchNextResults.autoplay.autoplay.replayVideoRenderer
            autoplay: {
                autoplay: {
                    sets: [
                        {
                            nextVideoRenderer: {
                                maybeHistoryEndpointRenderer: nextVideoItem,
                                autoplayEndpointRenderer: nextVideoItem
                            }
                        }
                    ],
                    replayVideoRenderer: itemWrapper
                }
            }
        }
    },
    // ButtonStates
    // $.transportControls.transportControlsRenderer
    transportControls: {
        transportControlsRenderer: buttonState
    }
};

JSON.stringify(watchNextResult);